api/rest/restapi.go

```diff
@@ -362,6 +363,93 @@ func (api *API) graphHandler(w http.ResponseWriter, r *http.Request) {
 	sendResponse(w, err, graph)
 }
 
+func (api *API) addFile(ctx context.Context, outChan <-chan *ipld.Node, w http.ResponseWriter) error {
+	for nodePtr := range outChan {
+		node := *nodePtr
+		/* Send block data to ipfs */
+		var hash string
+		/*		c := node.Cid()
+				format, ok := cid.CodecToStr[c.Type()]
+				if !ok {
+					format = ""
+				}*/
+		b := types.BlockWithFormat{
+			Data:   node.RawData(),
+			Format: "",
+		}
+		err := api.rpcClient.Call("",
+			"Cluster",
+			"IPFSBlockPut",
+			b,
+			&hash)
+		/* Verify that block put cid matches */
+		if node.String() != hash { // node string is just cid string
+			logger.Warningf("mismatch. node cid: %s\nrpc cid: %s", node.String(), hash)
+		}
+		if err != nil {
+			// TODO: think about how to handle this better.
+			// We may not want to stop all work after one failure
+			logger.Error(err)
+			sendAcceptedResponse(w, errors.New("error forwarding block"))
+			return err
+		}
+	}
+
+	// TODO: when complete this call should answer with a cid
+	// or better yet a pin-info describing the allocations
+	// of the resulting allocation
+	//
+	// Before returning this we will need to trigger a pin
+	// probably doing the same thing as Pin if its not a sharding call
+	//
+	// This is actually a bit tricky because right now we are not
+	// recording the root of the imported file, we could have ToChannel
+	// provide a root channel that can be read once the other channel is
+	// closed that reports the root at the end.  Probably would need to
+	// rewrite add for this to work
+	return nil
+}
+
+func (api *API) addShardedFile(ctx context.Context, outChan <-chan *ipld.Node, w http.ResponseWriter) error {
+	for nodePtr := range outChan {
+		node := *nodePtr
+		nodeS := types.NodeSerial{
+			CidS: node.Cid().String(),
+			Data: node.RawData(),
+		}
+		err := api.rpcClient.Call("",
+			"Cluster",
+			"ShardAddNode",
+			nodeS,
+			&struct{}{})
```

Comment:
Yeah I want to do this soon.  At the very least just the root hash of the final cluster DAG node (file root if adding locally) so that the user gets some closure.  The silent finish is kind of unnerving right now (what happened, did everything work correctly?).

Reply:
Latest commits address in large part, though concurrent request reading and response streaming is still elusive.